File: packages/leetcode/src/44.通配符匹配.ts
/*
 * @lc app=leetcode.cn id=44 lang=typescript
 *
 * [44] 通配符匹配
 */

// @lc code=start

/**
 * 初始化动态规划数组
 * @param s 待匹配的字符串
 * @param p 通配符模式
 * @returns [字符串长度, 模式长度, dp数组]
 */
function init(s: string, p: string) {
    const sLength = s.length;
    const pLength = p.length;

    const dp = Array.from({ length: sLength + 1 }, () => new Array(pLength + 1).fill(false)) as Array<Array<boolean>>;

    dp[0][0] = true;

    return [sLength, pLength, dp] as const;
}

function isMatch(s: string, p: string): boolean {
    const [sLength, pLength, dp] = init(s, p);

    for (let pIndex = 0; pIndex < pLength; pIndex++) {
        for(let sIndex = 0; sIndex < sLength; sIndex++) {
            
        }
    }

    return dp[sLength][pLength];
};
// @lc code=end

File: packages/leetcode/src/44.通配符匹配.go
/*
 * @lc app=leetcode.cn id=44 lang=golang
 *
 * [44] 通配符匹配
 */

// @lc code=start
package main

func isMatch(s string, p string) bool {
	m, n := len(s), len(p)
	dp := make([][]bool, m+1)
	for i := 0; i <= m; i++ {
		dp[i] = make([]bool, n+1)
	}
	dp[0][0] = true

	// 先处理掉第 dp[0][x] 的内容

	// 开始进行遍历
	for i := 0; i <= m; i++ {
		var curS rune
		if i != 0 {
			curS = rune(s[i-1])
		}

		for j := 1; j <= n; j++ {
			curP := rune(p[j-1])

			// 如果是*，则只需要看前面是否匹配正确即可
			if curP == '*' {
				if i == 0 {
					dp[i][j] = dp[0][j - 1]
					continue
				}

				if dp[i-1][j-1] == true || dp[i][j-1] == true || dp[i-1][j] {
					dp[i][j] = true
				} else {
					dp[i][j] = false
				}
				continue
			}

			if i == 0 {
				dp[i][j] = false
				continue
			}

			// ?一定要匹配一个字符，所以需要判断前面 i-1 个长度的 s 和前面 j-1 个长度的 p 是否匹配
			if curP == '?' {
				if dp[i-1][j-1] == true {
					dp[i][j] = true
				} else {
					dp[i][j] = false
				}

				continue
			}

			if curS == curP {
				dp[i][j] = dp[i-1][j-1]
			} else {
				dp[i][j] = false
			}
		}
	}

	return dp[m][n]
}

// @lc code=end
